Runtime/Core/ConfinerOven.cs

```diff
@@ -0,0 +1,307 @@
+        private ConfinerState ConfinerStateLerp(in ConfinerState left, in ConfinerState right, float frustumHeight)
+        {
+            if (left.m_Polygons.Count != right.m_Polygons.Count)
+            {
+                Assert.IsTrue(false, "Error in ConfinerStateLerp - Let us know on the Cinemachine forum please!");
+                return left;
+            }
+            ConfinerState result = new ConfinerState
+            {
+                m_Polygons = new List<ShrinkablePolygon>(left.m_Polygons.Count),
+            };
+            
+            float lerpValue = Mathf.InverseLerp(left.m_FrustumHeight, right.m_FrustumHeight, frustumHeight);
+            for (int i = 0; i < left.m_Polygons.Count; ++i)
+            {
+                var r = new ShrinkablePolygon(
+                    left.m_Polygons[i].m_AspectRatio,
+                    left.m_Polygons[i].m_AspectRatioBasedDiagonal,
+                    left.m_Polygons[i].m_NormalDirections)
+                {
+                    m_Points = new List<ShrinkablePolygon.ShrinkablePoint2>(left.m_Polygons[i].m_Points.Count),
+                    m_FrustumHeight = frustumHeight,
+                };
+                for (int j = 0; j < left.m_Polygons[i].m_Points.Count; ++j)
+                {
+                    r.m_IntersectionPoints = left.m_Polygons[i].m_IntersectionPoints;
+                    Vector2 rightPoint = right.m_Polygons[i].ClosestPolygonPoint(left.m_Polygons[i].m_Points[j]);
+                    r.m_Points.Add(new ShrinkablePolygon.ShrinkablePoint2
+                    {
+                        m_Position = Vector2.Lerp(left.m_Polygons[i].m_Points[j].m_Position, rightPoint, lerpValue),
+                        m_OriginalPosition = left.m_Polygons[i].m_Points[j].m_OriginalPosition,
+                    });
+                }
+                result.m_Polygons.Add(r);   
+            }
+            return result;
+        }
+        
+        private List<ConfinerState> m_confinerStates;
+        /// <summary>
+        /// Converts and returns m_shrinkablePolygons into List<ConfinerState>
+        /// </summary>
+        public List<ConfinerState> GetShrinkablePolygonsAsConfinerStates()
+        {
+            TrimShrinkablePolygons();
+
+            m_confinerStates = new List<ConfinerState>();
+            for (int i = 0; i < m_shrinkablePolygons.Count; ++i)
+            {
+                float stateAverage = m_shrinkablePolygons[i].Count;
+                for (int j = 0; j < m_shrinkablePolygons[i].Count; ++j)
+                {
+                    stateAverage += m_shrinkablePolygons[i][j].m_State;
+                }
+                stateAverage /= m_shrinkablePolygons[i].Count + 1;
+                
+                m_confinerStates.Add(new ConfinerState
+                {
+                    m_FrustumHeight = m_shrinkablePolygons[i][0].m_FrustumHeight,
+                    m_Polygons = m_shrinkablePolygons[i],
+                    m_State = stateAverage,
```

Comment:
this state averaging business looks a little iffy to me.  It's probably ok if the states change only monotonically.  Is that the case?

Reply:
Yes, they change monotonically.

Reply:
Not anymore, since my penalties are added. So I fixed this part too.